Mathlib/Tactic/Says.lean

```diff
@@ -0,0 +1,88 @@
+/-
+Copyright (c) 2023 Kim Liesinger. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kim Liesinger
+-/
+import Std.Data.String.Basic
+import Std.Tactic.GuardMsgs
+
+/-!
+# The `says` tactic combinator.
+
+If you write `X says`, where `X` is a tactic that produces a "Try this: Y" message,
+then you will get a message "Try this: X says Y".
+Once you've clicked to replace `X says` with `X says Y`,
+afterwards `X says Y` will only run `Y`.
+
+The typical usage case is:
+```
+simp? [X] says simp only [X, Y, Z]
+```
+-/
+
+open Lean Elab Tactic
+open Std.Tactic.TryThis
+
+namespace Mathlib.Tactic.Says
+
+register_option says.verify : Bool :=
+  { defValue := false
+    group := "says"
+    descr := "For every appearance of the `X says Y` combinator," ++
+      " re-verify that running `X` produces `Try this: Y`." }
+
+/-- Run `evalTactic`, capturing any new messages.-/
+def evalTacticCapturingMessages (tac : TSyntax `tactic) : TacticM MessageLog := do
+  let initMsgs ← modifyGetThe Core.State fun st => (st.messages, { st with messages := {} })
+  evalTactic tac
+  let msgs := (← getThe Core.State).messages
+  modifyThe Core.State fun st => { st with messages := initMsgs }
+  return msgs
+
+open Parser Tactic
+
+/-- This is a slight modification of `Parser.runParserCategory`. -/
+def parseAsTacticSeq (env : Environment) (input : String) (fileName := "<input>") :
+    Except String (TSyntax ``tacticSeq) :=
+  let p := andthenFn whitespace Tactic.tacticSeq.fn
+  let ictx := mkInputContext input fileName
+  let s := p.run ictx { env, options := {} } (getTokenTable env) (mkParserState input)
+  if s.hasError then
+    Except.error (s.toErrorMsg ictx)
+  else if input.atEnd s.pos then
+    Except.ok ⟨s.stxStack.back⟩
+  else
+    Except.error ((s.mkError "end of input").toErrorMsg ictx)
+
+syntax (name := says) tactic " says" (tacticSeq)? : tactic
```

Comment:
Have you checked that if you do
```
  simp says
  exact foo
```
that the parser doesn't consume the `exact foo` as well? If it does, and if there's no way to require the `tacticSeq` has strictly more indentation than the first tactic, then maybe we should alter the notation to be `tactic " says" ("(" tacticSeq ")")?` or something similar? Anything that serves to delimit the quotation for what's being said.

Reply:
the normal delimitation here would be indentation

Reply:
Add a `colGt` to the syntax suffices to deal with non-indented following tactics, happily! I've added some tests to exercise this.